src/stash/src/upgrade/v33_to_v34.rs

```diff
@@ -94,124 +100,140 @@ async fn migrate_cluster_replicas(tx: &mut Transaction<'_>) -> Result<(), StashE
 /// consumers of the audit log that might have stashed old events at some external place.
-async fn migrate_audit_log(tx: &mut Transaction<'_>, now: u64) -> Result<(), StashError> {
-    // We need to allocate additional audit events, which need IDs.
-    // Gather the current next_id.
+async fn migrate_audit_log(
+    tx: &mut Transaction<'_>,
+    entries: Vec<(u64, v34::ClusterReplicaKey, v34::ClusterReplicaValue)>,
+    now: u64,
+) -> Result<(), StashError> {
+    // Read in all of the active clusters so we can get the name for the audit log.
+    let clusters = tx
+        .peek_one::<v33::ClusterKey, v33::ClusterValue>(
+            tx.collection(CLUSTER_COLLECTION.name()).await?,
+        )
+        .await?;
+
+    // We need to allocate additional audit events, which need IDs. Gather the current next_id.
     let coll = ID_ALLOCATOR_COLLECTION.from_tx(tx).await?;
     let key = v33::IdAllocKey {
         name: AUDIT_LOG_ID_ALLOC_KEY.to_string(),
     };
     let mut next_audit_log_id = tx.peek_key_one(coll, &key).await?.unwrap().next_id;
 
-    // Collect all existing audit log events.
-    let coll = AUDIT_LOG_COLLECTION.from_tx(tx).await?;
-    let audit_events = tx.peek_one(coll).await?;
-    let mut events: Vec<_> = audit_events
-        .into_keys()
-        .map(|key| {
-            let v33::audit_log_key::Event::V1(event) = key.event.unwrap();
-            event
-        })
-        .collect();
-
-    // Find live replicas by iterating through the events in order and keeping track of the
-    // set of replicas not yet dropped. When every event is processed, this set contains
-    // only the currently live replicas.
-    events.sort_by_key(|e| e.id);
-    let mut live_replicas = BTreeMap::new();
-    for event in events {
-        let details = event.details.unwrap();
-        match details {
-            v33::audit_log_event_v1::Details::CreateClusterReplicaV1(details) => {
-                let key = (details.cluster_id.clone(), details.replica_id.clone());
-                let prev = live_replicas.insert(key, details);
-                assert_eq!(prev, None, "replica created twice");
-            }
-            v33::audit_log_event_v1::Details::DropClusterReplicaV1(details) => {
-                let key = (details.cluster_id, details.replica_id);
-                let prev = live_replicas.remove(&key);
-                assert!(prev.is_some(), "non-existing replica dropped");
-            }
-            _ => (),
+    let mut new_events = Vec::with_capacity(entries.len());
+    for (old_id, new_key, new_value) in entries {
+        let cluster_id = match new_value.cluster_id {
+            Some(v34::ClusterId {
+                value: Some(v34::cluster_id::Value::User(id)),
+            }) => format!("u{id}"),
+            Some(v34::ClusterId {
+                value: Some(v34::cluster_id::Value::System(id)),
+            }) => format!("s{id}"),
+            Some(v34::ClusterId { value: None }) | None => "".to_string(),
```

Comment:
Just for my understanding: This last case (and the `None` `ReplicaId` case below) should never happen, right? We are just handling them here to be as robust as possible?